server/src/test/java/org/elasticsearch/action/admin/indices/shrink/ResizeRequestTests.java

```diff
@@ -92,11 +105,79 @@ public void testToAndFromXContent() throws IOException {
     }
 
+    public void testEqualsAndHashCode() {
+        ResizeRequest request = createTestItem();
+        assertEquals(request, request);
+        assertEquals(request.hashCode(), request.hashCode());
+
+        ResizeRequest copy = new ResizeRequest();
+        copy.setTargetIndex(request.getTargetIndexRequest());
+        copy.setSourceIndex(request.getSourceIndex());
+        copy.setResizeType(request.getResizeType());
+        copy.setCopySettings(request.getCopySettings());
+        copy.setMaxSinglePrimarySize(request.getMaxSinglePrimarySize());
+        assertEquals(request, copy);
+        assertEquals(copy, request);
+        assertEquals(request.hashCode(), copy.hashCode());
+
+        // Changing targetIndexRequest makes requests not equal
+        copy.setTargetIndex(new CreateIndexRequest(request.getTargetIndexRequest().index() + randomAlphaOfLength(1)));
+        assertNotEquals(request, copy);
+        assertNotEquals(request.hashCode(), copy.hashCode());
+
+        // Changing sourceIndex makes requests not equal
+        copy.setSourceIndex(request.getSourceIndex() + randomAlphaOfLength(1));
+        assertNotEquals(request, copy);
+        assertNotEquals(request.hashCode(), copy.hashCode());
+
+        // Changing resizeType makes requests not equal
+        copy.setResizeType(ResizeType.SPLIT);
+        assertNotEquals(request, copy);
+        assertNotEquals(request.hashCode(), copy.hashCode());
+
+        // Changing resizeType makes requests not equal
+        if (request.getCopySettings()) {
+            copy.setCopySettings(null);
+        } else {
+            copy.setCopySettings(true);
+        }
+        assertNotEquals(request, copy);
+        assertNotEquals(request.hashCode(), copy.hashCode());
+
+        // Changing maxSinglePrimarySize makes requests not equal
+        if (request.getMaxSinglePrimarySize() != null) {
+            copy.setMaxSinglePrimarySize(null);
+        } else {
+            copy.setMaxSinglePrimarySize(new ByteSizeValue(randomIntBetween(1, 100)));
+        }
+        assertNotEquals(request, copy);
+        assertNotEquals(request.hashCode(), copy.hashCode());
+    }
+
+    public void testSerializeRequest() throws IOException {
+        ResizeRequest request = createTestItem();
+        BytesStreamOutput out = new BytesStreamOutput();
+        request.writeTo(out);
+        BytesReference bytes = out.bytes();
+        NamedWriteableRegistry namedWriteableRegistry = new NamedWriteableRegistry(NetworkModule.getNamedWriteables());
+        StreamInput wrap = new NamedWriteableAwareStreamInput(bytes.streamInput(),
+            namedWriteableRegistry);
+        ResizeRequest deserializedReq = new ResizeRequest(wrap);
+
+        assertEquals(request.getSourceIndex(), deserializedReq.getSourceIndex());
+        assertEquals(request.getTargetIndexRequest().settings(), deserializedReq.getTargetIndexRequest().settings());
```

Comment:
```
assertEquals(request.getTargetIndexRequest(), deserializedReq.getTargetIndexRequest());
```

Adding ^ here would make `testSerializeRequest()` fail, and so I think that means the `equals` method isn't actually correct as written (well, the code of it is fine, but since `CreateIndexRequest` doesn't override `equals`, too, we still end up doing the comparison as identity rather than equivalence there.

Reply:
I'm still looking into this -- it's a bit of a sticky situation. 😬

Reply:
Can we just compare the `settings` and `alias` of the `targetIndexRequest` in the `equals` method?

Reply:
I pulled some of this out into #68793 so that we can get it merged through before 7.12.0, but I'm still fussing with things on the `equals`/`hashCode` front. Thanks for your patience, @gaobinlong.